### ISSUE TITLE
libsodium.js not working anymore on iOS since 11.2.2 due to meltdown/spectre

### ISSUE THREAD

**Issue by tpetry** [jedisct1/libsodium.js#139]
At the current time, libsodium.js on iOS 11.2.2 up to 11.2.6 does not work. The functions throw an "Out of bounds memory access" error:
```
Error: Out of bounds memory access (evaluating 'R.asm._free.apply(null,arguments)')
wasm function: 31
wasm function
_free
_free — libsodium-sumo.min.js:5:279071
l — libsodium-wrappers.min.js:5:2597
d — libsodium-wrappers.min.js:5:2648
ut — libsodium-wrappers.min.js:5:45636
(anonyme Funktion) — show:61
promiseReactionJob
```

The following example works wonderfully in desktop browsers. However, on an iOS device it throws the mentioned errors:
https://jsfiddle.net/mouLbacd/1/

The error is already discussed in [kripken/emscripten#6042](https://github.com/kripken/emscripten/issues/6042). It seems as if it's on wasm and recompiling with asm.js could solve the problem. But the problem is still being discussed. The fact is that libsodium.js in the current state is defective on all iOS devices since 11.2.2.

**Comment by jedisct1** [comment]
Not sure that this is directly related to Spectre mitigations, but that definitely sounds like a regression. As if Apple didn't screw up enough already.

Great :(

**Comment by derrohrbach** [comment]
Yup! I'm running into exactly the same problem on iOS. Is there a way we can manually switch to the asm.js version of libsodium on devices where the wasm fails? We could use a switch/flag/function like that in conjuction with a detection code snippet from the other thread to circumvent the error on faulty iOS versions.

The detection code from the other thread: [https://github.com/brion/min-wasm-fail/blob/master/min-wasm-fail.js](https://github.com/brion/min-wasm-fail/blob/master/min-wasm-fail.js)

**Comment by jedisct1** [comment]
The code deciding to use asmjs or wasm is generated by Emscripten. Working around this looks complicated.

**Comment by derrohrbach** [comment]
What can i do now? My whole app depends on libsodium.js. I need to somehow work around this and it seems like the emscripten team does not want to work around it.

**Comment by jedisct1** [comment]
Fix it?

Try to understand how emscripten does wasm detection. There might be a property that can be redefined to make the wasm check fail.

**Comment by jedisct1** [comment]
Also, report it on Apple developer forums, or Apple is unlikely to fix it in future releases.

**Comment by buu700** [comment]
>The code deciding to use asmjs or wasm is generated by Emscripten. Working around this looks complicated.

Actually, that's in libsodium. This could be mitigated with something to the effect of `if ($IS_BROKEN_VERSION_OF_IOS) { reject(); return; }` right [here](https://github.com/jedisct1/libsodium/blob/029652e19c0f973da1581685830e9cab30136a04/dist-build/emscripten.sh#L100).

(I wouldn't mind having that in emscripten, or in some higher-level tool / shell script that factored out the logic that I pretty much copied and pasted between @cyph's wasm libs and libsodium, but as-is emscripten's only contribution is to fail at run time when WebAssembly is unsupported.)

**Comment by derrohrbach** [comment]
Awesome! Thank you @buu700!

**Comment by tpetry** [comment]
Could you please update the builds in the dist folder to use the new version? We tried now for some time building the libsodium.js files but have failed all the time.

**Comment by jedisct1** [comment]
I tried to update the builds today, but emscripten seems to have some issues: https://github.com/kripken/emscripten/issues/6387

**Comment by tpetry** [comment]
This has been fixed with iOS 11.3, nonetheless a new release with a fixed version for 11.2 would be helpful

**Comment by jedisct1** [comment]
@tpetry Help me fix the issue above and you'll have it :)

**Comment by catsigma** [comment]
Need a real fix for this, or please add the asm.js version in /dist.

**Comment by ryancwalsh** [comment]
@tpetry @catsigma Do you know of any workarounds for getting libsodium-js to work with iOS 11.2? 

My Apple devices are too old to upgrade beyond 11.2, and I'd love for libsodium-js to work.  Thanks.

**Comment by buu700** [comment]
Can you guys all confirm that the latest libsodium.js still isn't working with iOS 11.2? I only have an iOS 11.4 device to test with, but can't think of any reason why the fix from a few months ago wouldn't work.

**Comment by ryancwalsh** [comment]
@buu700 I'm basing this both on my own app and also this demo test case from the top of this issue: https://jsfiddle.net/mouLbacd/1/ That fiddle works fine on my desktop but shows the error on my iPhone 6 with iOS 11.2. I assume it would be pulling from the latest libsodium-js code. Thanks.

**Comment by buu700** [comment]
Ah, it looks like that's using the latest npm package, which hasn't been updated in 8 months. (Is dynamic memory growth support what that's blocked on, @jedisct1? If so, maybe the best solution for now would be to bump the major version, mark that feature as no longer supported, and throw an informative error whenever someone tries to use it. Also, might be worth running new `ALLOW_MEMORY_GROWTH` benchmarks against [the latest V8](https://v8project.blogspot.com/2018/08/liftoff.html).)

Anyway, can you still reproduce the problem with [this jsfiddle](https://jsfiddle.net/mouLbacd/3)?

**Comment by buu700** [comment]
Hm, what about [this one](https://jsfiddle.net/mouLbacd/6)?

**Comment by ryancwalsh** [comment]
@buu700 That one seems to work. It alerts "32" on both iOS and desktop.

**Comment by buu700** [comment]
Interesting, thanks. I see the general issue with our existing fix then; I'll take a stab at resolving it probably some time today.

**Comment by ryancwalsh** [comment]
@buu700 Thanks for looking into this. I also just tried https://codesandbox.io/s/4ryxr5xyz0?autoresize=1&expanddevtools=1&hidenavigation=1, which works on Windows 10 Chrome desktop but causes this error on iOS 11.2.6:

```
Unhandled Rejection (Error)
abort("Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 16777216, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 ") at t
a
N
f
enlargeMemory
wasm function@[wasm code]
wasm function: 46@[wasm code]
```

**Comment by jedisct1** [comment]
Can you check the current builds (in the `dist` directory, not on npm)?

**Comment by ryancwalsh** [comment]
@jedisct1 I'm not sure whether I'm correctly doing what you're suggesting. 

I visited https://github.com/jedisct1/libsodium.js/tree/9c7f4cd571f9a7cf1ff7648aea489b5b5417a13e/dist/modules and noticed that 'libsodium-wrappers.js' hasn't been updated recently, so I only downloaded libsodium.js.

My code in the past has been using `const _sodium = require('libsodium-wrappers');`, which I now changed to `const _sodium = require('libsodium');` (and I similarly updated my `webpack.mix.js`). 

In my project's `node_modules/libsodium/dist/modules/`, I overwrote the `libsodium.js` file that was there with the one that I just downloaded.

`npm run production` seems to work, but in my desktop browser console, I now see this warning: 

> jQuery.Deferred exception: o.from_hex is not a function H9cq/i</</t.decrypt@https:...

And decryption of course doesn't work. I haven't tested iPhone because I assume that's broken now too.

I also assume that there is a relevance between 'libsodium-wrappers' and 'libsodium' and that this probably isn't what you meant for me to do.

I'll eagerly try whatever you suggest next.

Thanks for looking into this.

cc @buu700

**Comment by buu700** [comment]
I don't know about the other issues, but it should at least work if you use https://github.com/jedisct1/libsodium.js/blob/master/dist/browsers-sumo/sodium.js.

**Comment by ryancwalsh** [comment]
I'm still not sure I'm understanding what to do.

This time, in my project's `node_modules/libsodium/dist/modules/`, I overwrote the libsodium.js file that was there with https://github.com/jedisct1/libsodium.js/blob/master/dist/browsers-sumo/sodium.js.

Then I kept my code as pointing to 'libsodium' instead of 'libsodium-wrappers'.

`npm run production` gave me this error:

```
ERROR in ./node_modules/libsodium/dist/modules/libsodium.js
Module not found: Error: Can't resolve 'libsodium-sumo' in 'C:\code\myapp\node_modules\libsodium\dist\modules'
 @ ./node_modules/libsodium/dist/modules/libsodium.js 1:719491-719529
```

So then in my code, I changed it to `require('libsodium-sumo')`. Then `npm run production` gave me this error:

```
ERROR in multi jquery bootstrap libsodium-sumo
Module not found: Error: Can't resolve 'libsodium-sumo' 
```

So I ran `npm install libsodium-sumo --save` and then `npm run production` worked, but my Chrome desktop browser still shows this warning:

> jQuery.Deferred exception: o.from_hex is not a function

Then I realized that what I just did probably made no sense. (You can tell I'm pretty confused about what all these files are and where they are supposed to go. I don't understand what 'sumo' is and 'browser' and the regular files and when/where to use each.)

I also saw https://github.com/jedisct1/libsodium.js/issues/140

So now I'm thinking that (for testing purposes at least) you want me to remove all mentions of libsodium _anything_ from my `package.json` and do a fresh `npm install`.

Then in my `webpack.mix.js`, I changed it to `"libsodium-browsers-sumo": path.resolve(__dirname, "resources/assets/js/vendor/sodium.js"),`, which is where I've now downloaded https://github.com/jedisct1/libsodium.js/blob/master/dist/browsers-sumo/sodium.js.

And in my code, I changed it to `require('libsodium-browsers-sumo')`.

Now compiling produces these errors:

```
ERROR in ./resources/assets/js/vendor/sodium.js
Module not found: Error: Can't resolve 'fs' in 'C:\code\myapp\resources\assets\js\vendor'
 @ ./resources/assets/js/vendor/sodium.js 1:1366-1379 1:388920-388933
...

ERROR in ./resources/assets/js/vendor/sodium.js
Module not found: Error: Can't resolve 'libsodium-sumo' in 'C:\code\myapp\resources\assets\js\vendor'
```

**Comment by ryancwalsh** [comment]
Okay, so the code I've always been using has been fine and is what I'll continue to use even for testing purposes here.

```
const _sodium = require('libsodium-wrappers');

    exportsSelf.sodiumReady = new $.Deferred();
    $.when(_sodium.ready).then(function (data, textStatus, jqXHR) {//https://github.com/jedisct1/libsodium.js
        const sodium = _sodium;
        console.log('libsodium-wrappers is ready');
        ...
```

But which file do you want me to download from the master branch in the repo, and where specifically should I put it in my project, and how should I "include" it (or will that `require('libsodium-wrappers')`) statement handle everything and pull in the new file somehow?

**Comment by buu700** [comment]
I was about to suggest trying [the jsfiddle from above](https://jsfiddle.net/mouLbacd/3) (which references dist/browsers-sumo/sodium.js from master), but it seems that something's gone terribly wrong based on my testing in desktop Chrome:

* The first time `sodium.crypto_aead_chacha20poly1305_keygen()` is run it returns all zeroes, then seems to work as expected afterwards.

* I got `sodium.crypto_secretbox_easy()` to return something that looked like valid input once, but most of the time I get `RangeError: Source is too large` and  sometimes I get `RangeError: offset is out of bounds`.

* `sodium.crypto_secretbox_open_easy()` consistently fails with `ReferenceError: var_size is not defined`.

**Comment by jedisct1** [comment]
When people report that everything appears to be working :)

Which doesn't seem to be the case yet, as Ryan reported a regression that still has to be figured out.

**Comment by buu700** [comment]
Hm, well would you be able to rebuild it again @jedisct1 for @ryancwalsh to test?

I just tried again, and after switching from Debian Stretch to OS X (`./configure` doesn't seem to like that the former has an outdated version of llvm) it fails with `crypto_secretstream/xchacha20poly1305/secretstream_xchacha20poly1305.c:126:5: error: array size is negative`.

**Comment by NullVoxPopuli** [comment]
ok @jedisct1 I found that if I revert to 8fa117918df38f52668feaf9cbc498c633671855 my build passes. :-\

**Comment by NullVoxPopuli** [comment]
actually, nevermind, it only passed once. :-\

**Comment by ryancwalsh** [comment]
@jedisct1 @buu700 Does this issue being closed mean "Libsodium-js won't support iOS 11.2 or Android 5.1. Use something else instead." ?

**Comment by buu700** [comment]
Everything should actually be good to go after https://github.com/jedisct1/libsodium/pull/763 is merged. Until then, to test it out @ryancwalsh, try [jsfiddle.net/8dtgkv6a](https://jsfiddle.net/8dtgkv6a).

**Comment by ryancwalsh** [comment]
@buu700 On iOS 11.2 Chrome and on Windows 10 Chrome 69, I get an alert box saying "test test". I get no alert box on Android 5.1, even when I changed your fiddle to https://jsfiddle.net/8dtgkv6a/2/

Thanks.

**Comment by buu700** [comment]
Hm, are you saying that even the `alert('About to start...')` you added isn't working?

**Comment by ryancwalsh** [comment]
@buu700 That doesn't trigger an alert either. And so I figured I'd try a different fiddle platform (ideally one that would show the console log on the actual device screen). So I created https://codesandbox.io/embed/8lxmkr5l9?autoresize=1&expanddevtools=1&hidenavigation=1&codemirror=1. But that one doesn't even load at all on my Android! Just a blank white page. 

So then I created https://jsbin.com/xeheqahefa/edit?html,js,console, but on Android 5.1, it produces `SyntaxError: Unexpected token )` in the console.  It seems likely that my Android is just so outdated that it can't handle much at all.

**Comment by ryancwalsh** [comment]
@buu700 I thought that changing my `package.json` from `"libsodium-wrappers": "^0.7.3",` to `"libsodium-wrappers": "git://github.com/jedisct1/libsodium.js.git#9d4455cfd8c68a53238475b9f900f42582d7c925",` would allow me to use [this commit](https://github.com/jedisct1/libsodium.js/commit/9d4455cfd8c68a53238475b9f900f42582d7c925) even before the official release.

But that works on Windows Chrome but not iOS 11.2. On iOS, I tried Safari, Firefox, and DuckDuckGo. It's weird to me that your https://jsfiddle.net/8dtgkv6a works on *some* iOS browsers but that 9d4455cfd8c68a53238475b9f900f42582d7c925 doesn't.

P.S. I tested `"libsodium-wrappers": "^0.7.3"` on an even older iOS, and it worked: iOS 10.3.3 (14G60) iPhone 6S.

**Comment by buu700** [comment]
Hmm, I wonder if those failures were related to libsodium itself or just those sites not supporting Android 5.1. Did the jsfiddle or anything else you tried work on iOS 11.2?

Also, what about starting a chat at https://cyph.im between two Safari tabs in iOS 11.2? (That uses the latest libsodium.js, so should fail if this fix doesn't work.)

**Comment by ryancwalsh** [comment]
@buu700 Sorry, I no longer have a device with iOS 11.2.

I do have an old iPad with iOS 9.1, but cyph.im never got beyond the welcome screen with the castle logo building.
